services/ansibler/server/loadbalancers.go

```diff
@@ -40,106 +40,316 @@ const (
+		DesiredLbCluster *pb.LBcluster
+	}
+
+	NginxPlaybookData struct {
+		Loadbalancer string
+	}
+
+	ConfData struct {
+		Roles []LBConfiguration
+	}
+
+	LBConfiguration struct {
+		Role        *pb.Role
+		TargetNodes []*pb.Node
+	}
+)
+
+// APIEndpointState determines if the API endpoint should be updated with a new
+// address, as otherwise communication with the cluster wouldn't be possible.
+func (lb *LBData) APIEndpointState() APIEndpointChangeState {
+	if lb.CurrentLbCluster == nil && lb.DesiredLbCluster == nil {
+		return NoChange
+	}
+
+	if lb.CurrentLbCluster == nil && lb.DesiredLbCluster != nil {
+		return AttachingLoadBalancer
+	}
+
+	if lb.CurrentLbCluster != nil && lb.DesiredLbCluster == nil {
+		return DetachingLoadBalancer
+	}
+
+	if lb.CurrentLbCluster.Dns.Endpoint != lb.DesiredLbCluster.Dns.Endpoint {
+		return EndpointRenamed
+	}
 
-type ConfData struct {
-	Roles []LBConfiguration
+	// check if role changed.
+	var isAPIServer bool
+	for _, currentRoles := range lb.CurrentLbCluster.Roles {
+		if currentRoles.RoleType == pb.RoleType_ApiServer {
+			isAPIServer = true
+			break
+		}
+	}
+
+	for _, desiredRoles := range lb.DesiredLbCluster.Roles {
+		if desiredRoles.RoleType == pb.RoleType_ApiServer && !isAPIServer {
+			return RoleChanged
+		}
+	}
+
+	return NoChange
 }
 
-type LBConfiguration struct {
-	Role        *pb.Role
-	TargetNodes []*pb.Node
+// tearDownLoadBalancers will correctly destroy load-balancers including correctly selecting the new ApiServer if present.
+func teardownLoadBalancers(deleted map[string]*LBInfo) error {
+	var k8sGroup errgroup.Group
+
+	for k8sClusterName, deleted := range deleted {
+		k8sClusterName := k8sClusterName
+		deleted := deleted
+
+		k8sDirectory := filepath.Join(baseDirectory, outputDirectory, fmt.Sprintf("%s-%s-lbs", k8sClusterName, utils.CreateHash(4)))
+
+		if err := generateK8sBaseFiles(k8sDirectory, deleted); err != nil {
+			log.Error().Msgf("error while generating base directory for %s : %v", k8sClusterName, err)
+			continue
+		}
+
+		k8sGroup.Go(func() error {
+			var apiServer *LBData
+
+			for _, lb := range deleted.LbClusters {
+				for _, role := range lb.CurrentLbCluster.Roles {
+					if role.RoleType == pb.RoleType_ApiServer {
+						apiServer = lb
+					}
+				}
+			}
+
+			if err := HandleAPIEndpointChange(apiServer, deleted, k8sDirectory); err != nil {
+				return err
+			}
+
+			return os.RemoveAll(k8sDirectory)
+		})
+	}
+
+	return k8sGroup.Wait()
 }
 
-//setUpLoadbalancers will set up and verify the loadbalancer configuration including DNS
-//returns error if not successful, nil otherwise
+// setUpLoadbalancers will set up and verify the loadbalancer configuration including DNS
+// returns error if not successful, nil otherwise
 func setUpLoadbalancers(lbInfos map[string]*LBInfo) error {
 	var errGroupK8s errgroup.Group
 	//iterate through k8s clusters LBs
 	for k8sClusterName, lbInfo := range lbInfos {
-		func(lbInfo *LBInfo, k8sClusterName string) {
```

Comment:
What is the reason to change the function around the error group to `var := var` ?

Reply:
started with the `func(a, b)(c,d)` wrapping as it was at the start, each time I looked at the function took me a while to to read the code due to the nested function creation and indentation, I just created local variables instead which has the same effect as passing it as arguments to the function wrapper, tbh I don't really care about which to choose I can revert it back to the `func(a,b)(c,d)` wrapper.

Reply:
We used a wrapper every time we needed this logic, thus for consistency's sake, I would choose one and roll with it. I don't mind either.

Reply:
👍  I'll revert it back to the function wrapper since its used in other places